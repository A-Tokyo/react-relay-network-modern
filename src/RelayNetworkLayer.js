/* @flow */

import { Network } from 'relay-runtime';
import RelayRequest from './RelayRequest';
import fetchWithMiddleware from './fetchWithMiddleware';
import type {
  Middleware,
  FetchFunction,
  SubscribeFunction,
  RNLExecuteFunction,
} from './definition';

export default class RelayNetworkLayer {
  _middlewares: Middleware[];
  execute: RNLExecuteFunction;
  fetchFn: FetchFunction;
  subscribeFn: ?SubscribeFunction;

  constructor(middlewares: Middleware[] | Middleware, subscribeFn?: SubscribeFunction) {
    this._middlewares = Array.isArray(middlewares) ? middlewares : [middlewares];
    this.subscribeFn = subscribeFn;

    this.fetchFn = (operation, variables, cacheConfig, uploadables) => {
      const req = new RelayRequest(operation, variables, cacheConfig, uploadables);
      return fetchWithMiddleware(req, this._middlewares);
    };

    this.execute = Network.create(this.fetchFn, this.subscribeFn);
  }
}
