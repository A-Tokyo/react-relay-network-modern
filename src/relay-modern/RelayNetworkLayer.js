/* @flow */

import { Network } from 'relay-runtime';
import type { Middleware } from '../definition';

export type Variables = { [name: string]: $FlowFixMe };
export type ConcreteBatch = any;
export type CacheConfig = {
  force?: ?boolean,
  poll?: ?number,
  rerunParamExperimental?: ?any,
};
export type Disposable = { dispose(): void };
export type Uploadable = File | Blob;
export type UploadableMap = { [key: string]: Uploadable };
export type PayloadData = { [key: string]: mixed };
export type QueryPayload = {|
  data?: ?PayloadData,
  errors?: Array<any>,
  rerunVariables?: Variables,
|};
// this is workaround should be class from relay-runtime/network/RelayObservable.js
export type RelayObservable<T> = Promise<T>;
// Note: This should accept Subscribable<T> instead of RelayObservable<T>,
// however Flow cannot yet distinguish it from T.
export type ObservableFromValue<T> = RelayObservable<T> | Promise<T> | T;
export type FetchFunction = (
  operation: ConcreteBatch,
  variables: Variables,
  cacheConfig: CacheConfig,
  uploadables: ?UploadableMap
) => ObservableFromValue<QueryPayload>;
// See SubscribeFunction type declaration in relay-runtime/network/RelayNetworkTypes.js
export type SubscribeFunction = (
  operation: ConcreteBatch,
  variables: Variables,
  cacheConfig: CacheConfig,
  observer: any
) => RelayObservable<QueryPayload> | Disposable;

export default class RelayModernNetworkLayer {
  _middlewares: Middleware[];
  execute: (
    operation: ConcreteBatch,
    variables: Variables,
    cacheConfig: CacheConfig,
    uploadables?: ?UploadableMap
  ) => RelayObservable<QueryPayload>;

  constructor(middlewares: Middleware[] | Middleware, subscribeFn?: SubscribeFunction) {
    this._middlewares = Array.isArray(middlewares) ? middlewares : [middlewares];

    const fetchFn: FetchFunction = (operation, variables, cacheConfig, uploadables) => {
      return Promise.resolve({});
    };

    this.execute = Network.create(fetchFn, subscribeFn);
  }
}
