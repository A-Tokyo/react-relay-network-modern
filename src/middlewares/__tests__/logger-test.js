/* @flow */

import fetchMock from 'fetch-mock';
import RelayNetworkLayer from '../../RelayNetworkLayer';
import { mockReq } from '../../__mocks__/mockReq';
import loggerMiddleware from '../logger';

describe('middlewares/logger', () => {
  beforeEach(() => {
    fetchMock.restore();
  });

  it('measure request time for request', async () => {
    fetchMock.mock({
      matcher: '/graphql',
      response: {
        status: 200,
        body: { data: 'PAYLOAD' },
      },
      method: 'POST',
    });

    const logger = jest.fn();
    const rnl = new RelayNetworkLayer([
      loggerMiddleware({
        logger,
      }),
    ]);

    await mockReq('MyRequest').execute(rnl);
    expect(logger).toHaveBeenCalledTimes(2);
    expect(logger.mock.calls).toMatchSnapshot();
  });
});
