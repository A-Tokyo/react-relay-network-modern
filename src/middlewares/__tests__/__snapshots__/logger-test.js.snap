// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`middlewares/logger measure request time for request 1`] = `
Array [
  Array [
    "Run MyRequest",
    RelayRequest {
      "cacheConfig": Object {},
      "fetchOpts": Object {
        "body": "{\\"id\\":\\"MyRequest\\",\\"query\\":\\"\\",\\"variables\\":{}}",
        "headers": Object {},
        "method": "POST",
      },
      "id": "MyRequest",
      "operation": Object {
        "id": "MyRequest",
        "text": "",
      },
      "uploadables": undefined,
      "variables": Object {},
    },
  ],
  Array [
    "Done MyRequest in XXXms",
    Object {
      "query": "",
      "req": RelayRequest {
        "cacheConfig": Object {},
        "fetchOpts": Object {
          "body": "{\\"id\\":\\"MyRequest\\",\\"query\\":\\"\\",\\"variables\\":{}}",
          "headers": Object {},
          "method": "POST",
        },
        "id": "MyRequest",
        "operation": Object {
          "id": "MyRequest",
          "text": "",
        },
        "uploadables": undefined,
        "variables": Object {},
      },
      "res": RelayResponse {
        "_res": Body {
          "_abort": false,
          "_bytes": 18,
          "_raw": Array [
            Object {
              "data": Array [
                123,
                34,
                100,
                97,
                116,
                97,
                34,
                58,
                34,
                80,
                65,
                89,
                76,
                79,
                65,
                68,
                34,
                125,
              ],
              "type": "Buffer",
            },
          ],
          "body": Readable {
            "_events": Object {
              "data": [Function],
              "end": [Function],
              "error": [Function],
            },
            "_eventsCount": 3,
            "_maxListeners": undefined,
            "_readableState": ReadableState {
              "awaitDrain": 0,
              "buffer": BufferList {
                "head": null,
                "length": 0,
                "tail": null,
              },
              "decoder": null,
              "defaultEncoding": "utf8",
              "destroyed": false,
              "emittedReadable": false,
              "encoding": null,
              "endEmitted": true,
              "ended": true,
              "flowing": true,
              "highWaterMark": 16384,
              "length": 0,
              "needReadable": false,
              "objectMode": false,
              "pipes": null,
              "pipesCount": 0,
              "readableListening": false,
              "reading": false,
              "readingMore": false,
              "resumeScheduled": false,
              "sync": true,
            },
            "domain": null,
            "readable": false,
          },
          "bodyUsed": true,
          "headers": Headers {
            "_headers": Object {},
          },
          "ok": true,
          "size": 0,
          "status": 200,
          "statusText": "OK",
          "timeout": 0,
          "url": "/graphql",
        },
        "data": "PAYLOAD",
        "headers": Headers {
          "_headers": Object {},
        },
        "json": Object {
          "data": "PAYLOAD",
        },
        "ok": true,
        "status": 200,
        "url": "/graphql",
      },
      "variables": Object {},
    },
  ],
]
`;
